File: src/Container.cpp
#include "Container.h"
#include "stdlib.h"
#include "time.h"

void InclusionSort(Container &container)
{
    for (int i = 1; i < container.length; i++)
    {
        Figure cur = container.figures[i];
        int ind = i;
        while (ind > 0 && Area(container.figures[ind - 1]) >= Area(cur))
        {
            container.figures[ind] = container.figures[ind - 1];
            ind -= 1;
        }
        container.figures[ind] = cur;
    }
}

int FillContainer(Container &container, int argc, char* argv[])
{
    if (argc < 2 || argc > 3)
    {
        printf("%s", "Incorrect amount of arguments.");
        return 1;
    }
    container.length = atoi(argv[1]);
    if (argc == 3) {
        FILE* file;
        if (container.length < 0 || container.length > 20) {
            printf("%s", "Incorrect amount of figures for this test.");
            return 1;
        }
        file = fopen(argv[2], "rt");
        for (int i = 0; i < container.length; i += 1) {
            int type;
            fscanf(file, "%u", &type);
            if (type == 1) {
                container.figures[i].circle = Circle();
                container.figures[i].circle.input_file = file;
                ReadCircle(&container.figures[i].circle);
                container.figures[i].type = circleType;
            }
            else if (type == 2) {
                container.figures[i].rectangle = Rectangle();
                container.figures[i].rectangle.input_file = file;
                ReadRectangle(&container.figures[i].rectangle);
                container.figures[i].type = rectangleType;
            }
            else if (type == 3) {
                container.figures[i].triangle = Triangle();
                container.figures[i].triangle.input_file = file;
                ReadTriangle(&container.figures[i].triangle);
                container.figures[i].type = triangleType;
            }
            else {
                printf("%s%u", "Wrong type of figure at input file.", type);
                return 1;
            }
        }
        for (int i = 0; i < container.length; i += 1) {
            pthread_join(container.figures[i].thread, NULL);
        }
        fclose(file);
    }
    else if (argc == 2)
    {
        if (container.length < 20 || container.length > 100000)
        {
            printf("%s", "Incorrect amount of figures for this test.");
            return 1;
        }
        srand(time(0));
        for (int i = 0; i < container.length; i += 1) {
            int type = rand() % 3;
            if (type == 1) {
                container.figures[i].circle = Circle();
                pthread_create(&container.figures[i].thread, NULL, GenerateCircle, &container.figures[i].circle);
                container.figures[i].type = circleType;
            }
            else if (type == 2) {
                container.figures[i].rectangle = Rectangle();
                pthread_create(&container.figures[i].thread, NULL, GenerateRectangle, &container.figures[i].rectangle);
                container.figures[i].type = rectangleType;
            }
            else if (type == 3) {
                container.figures[i].triangle = Triangle();
                pthread_create(&container.figures[i].thread, NULL, GenerateTriangle, &container.figures[i].triangle);
                container.figures[i].type = triangleType;
            }
            pthread_join(container.figures[i].thread, NULL);
        }
        for (int i = 0; i < container.length; i += 1) {
            pthread_join(container.figures[i].thread, NULL);
        }
    }
    return 0;
}

File: Container.py
from Circle import Circle
from Rectangle import Rectangle
from Triangle import Triangle
import sys
import random
import time


class Container:
    def __init__(self):
        self.figures = []
        self.length = 0

    def inclusion_sort(self):
        for i in range(0, self.length):
            cur = self.figures[i]
            ind = i
            while ind > 0 and self.figures[ind - 1].area() >= cur.area():
                self.figures[ind] = self.figures[ind - 1]
                ind -= 1
            self.figures[ind] = cur

    def fill_container(self):
        argv = sys.argv
        argc = len(sys.argv)
        if argc < 2 or argc > 3:
            print("Incorrect amount of arguments.")
            return 1
        self.length = int(argv[1])
        if argc == 3:
            if self.length < 0 or self.length > 20:
                print("Incorrect amount of figures for this test.")
                return
            file = open(argv[2])
            for i in range(0, self.length):
                type = int(file.readline())
                if type == 1:
                    self.figures.append(Circle())
                    self.figures[-1].read(file)
                elif type == 2:
                    self.figures.append(Rectangle())
                    self.figures[-1].read(file)
                elif type == 3:
                    self.figures.append(Triangle())
                    self.figures[-1].read(file)
                else:
                    print("Wrong type of figure at input file.")
                    return 1
            file.close()
        elif argc == 2:
            if self.length < 20 or self.length > 100000:
                print("Incorrect amount of figures for this test.")
                return 1
            random.seed(time.time())
            for i in range(0, self.length):
                type = random.randint(0, 2)
                if type == 0:
                    self.figures.append(Circle())
                    self.figures[-1].generate()
                elif type == 1:
                    self.figures.append(Rectangle())
                    self.figures[-1].generate()
                elif type == 2:
                    self.figures.append(Triangle())
                    self.figures[-1].generate()
        return 0
